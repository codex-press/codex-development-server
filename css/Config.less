
.Config {
  position: fixed;
  width: 500px;
  height: 500px;
  left: calc(50% - 250px);
  top: calc(50% - 250px);
  padding: 1em;

  &:extend(.modal);


  label {
    display: block;
    margin: 1em 0;
  }

  label.invalid {
    color: red;
  }

  label.invalid input:not(:focus) {
    outline: red solid 1px;
  }

  label.invalid input {
    outline-color: red;
  }

  input[type=text] {
    width: 100%;
    font-size: 20px;
    outline-color: white;
    text-overflow: ellipsis;
  }

  button {
    font-size: 1.3rem;
    background-color: hsl(206, 83%, 42%);
    color: white;
    float: right;
    border: 0;
    border-radius: 0.1em;
    padding: 0.3em 0.8em;
    user-select: none;
    cursor: pointer;
  }


  button {
    outline-color: hsl(206, 83%, 42%);
  }

}

