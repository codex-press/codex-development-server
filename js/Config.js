import React, {Component} from 'react';

import ConfigModal from './ConfigModal';


export default class Config extends Component {

  render() {
    return (
      <div className="Config">

        <link rel="stylesheet" href="/main.css" />

        <div className="dimmer"></div>

        <ConfigModal />
      </div>
    );
  }

}


