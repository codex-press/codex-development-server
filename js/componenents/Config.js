import React from 'react';
import { connect } from 'react-redux'

import { addRepo, removeRepo, updateRepo, setToken } from '../actions';

import { api } from '../utility'

import { IconValid, IconInvalid } from './icons';

import Repository from './Repository';
import RepositoryAdd from './RepositoryAdd';


const mapStateToProps = state => {
  return {
    ...state.get('config').toObject(),
    open           : state.getIn(['ui','modal']) === 'config',
    repositoryList : state.get('repositories').toJS(),    
    account        : state.get('account') && state.get('account').toObject(),
  }
}


const mapDispatchToProps = {
  addRepo,
  removeRepo,
  updateRepo,
  setToken,
}


class ConfigModal extends React.Component {


  constructor() {
    super();
    this.state = { addRepo: false, };
  }


  render() {
    
    if (!this.props.open)
      return null;

    let homeUrl;
    if (this.props.account) {
      homeUrl = 'https://codex.press' + this.props.account.root_directory_url;
    }

    return (
      <div 
        className="Config"
        ref={ el => this.topEl = el }
        onWheel={ e => this.wheel(e) }>

        <h2>Codex Press Develepment Server</h2>

        { this.props.account && !this.state.changingLogin &&
          <div className="account-details">
            <div>Logged in as { this.props.account.name }</div>
            <div>{ this.props.account.email }</div>
            <div><a href={ homeUrl } target="_blank">{ homeUrl }</a></div>
            <button 
              className="small"
              onClick={ () => this.setState({changingLogin: true}) }>
              Change Account
            </button>
          </div>
        }

        { this.state.changingLogin &&
          <form 
            className={ this.state.invalidLogin ? 'invalid' : '' }
            onSubmit={ e => this.changeLogin(e) } >

            <label>
              Email
              <input type="email" />
            </label>

            <label>
              Password
              <input type="password" />
            </label>

            <button
              type="button"
              onClick={ () => this.setState({changingLogin: false}) } >
              Cancel
            </button>

            {' '}

            <button type="submit" className="primary">
              Change
            </button>

          </form>
        }

        <h3>Repositories</h3>

        <a href="https://codex.press/edit/repos" target="_blank">
          View on production
        </a>

        { this.props.repositories.size === 0 ?

          <div className="empty">No repositories!</div> :

          this.props.repositories.map((repo, name) => 
            <Repository
              key={ name }
              { ...repo.toJS() }
              remove={ () => this.props.removeRepo(name) }
              files={
                this.props.repositoryList[name] &&
                this.props.repositoryList[name].files ||
                []
              }
              change={ (newName, path) => {
                this.props.updateRepo(name, newName, path) 
              } }
              checkName={ name => this.checkName(name) }
            />
          ).toArray()
        }

        <RepositoryAdd
          submit={ (name, path) => this.props.addRepo(name, path) }
          checkName={ name => this.checkName(name) }
        />

      </div>
    );
  }


  changeLogin(e) {
    e.preventDefault();
    let email = e.target.querySelector('[type="email"]').value;
    let password = e.target.querySelector('[type="password"]').value;
    api('/login', {post: {email, password}, auth: false})
    .then(data => {
      this.setState({changingLogin: false})
      this.props.setToken(data.token);
    })
    .catch(data => this.setState({invalidLogin: true}));
  }


  wheel(e) {
    this.topEl.scrollTop += e.deltaY;
    e.preventDefault();
  }


  checkName(name) {
    return (
      name.length > 1 &&
      !/[^\w-]/.test(name) &&
      !this.props.repositories.has(name)
    );
  }

}



export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ConfigModal);


