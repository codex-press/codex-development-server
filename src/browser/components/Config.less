@import (reference) "../index.less";

.Config {
  position: fixed;
  width: 500px;
  height: 500px;
  left: calc(50% - 250px);
  top: calc(50% - 250px);
  padding: 1em;
  overflow: auto;

  &:extend(.modal);

  h2 {
    margin: 0 0 1rem 0;
    text-align: center;
  }

  button {
    margin: 0.5rem 0;
  }

  h3 {
    font-size: 22px;
    margin-bottom: 0;
  }

}

