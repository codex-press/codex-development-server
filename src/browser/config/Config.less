@import (reference) "../index.less";

.Config {
  padding: 1em;
  overflow: auto;

  &:extend(.modal);

  h2 {
    margin: 0 0 1rem 0;
    text-align: center;
  }

  button {
    margin: 0.5rem 0;
  }

  h3 {
    font-size: 1.5rem;
    margin-top: 3rem;
    margin-bottom: 0;
  }

  label {
    display: block;
    margin: 1rem 0;
  }
  
  label.for-checkbox {
    cursor: pointer;
  }

  input[type=checkbox] {
    vertical-align: top;
    cursor: pointer;
  }

  select {
    cursor: pointer;
  }

  .impersonate {
    margin: 1rem 0;
  }

}

