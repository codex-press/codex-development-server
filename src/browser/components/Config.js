import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux'

import {
  addRepository,
  removeRepository,
  changeRepository,
  setToken
} from '../actions';

import AccountDetails from './AccountDetails';
import Repository from './Repository';
import RepositoryAdd from './RepositoryAdd';

import './Config.less';

const mapStateToProps = state => {
  return {
    config       : state.get('config'),
    repositories : state.get('repositories'),
    account      : state.get('account'),
  }
}


const mapDispatchToProps = {
  addRepository,
  removeRepository,
  changeRepository,
  setToken,
}


export function Config(props) {

  const wheel = e => {
    e.preventDefault();
    e.target.scrollTop += e.deltaY;
  }

  const usedNames = props.repositories.map(a => a.get('name'));

  const changeRepository = (newName, path) => {
    props.changeRepository(name, newName, path) 
  }

  return (
    <div className="Config" onWheel={ wheel }>

      <h2>Codex Press Develepment Server</h2>

      <AccountDetails account={ props.account } />

      <h3>Repositories</h3>

      <a href="https://codex.press/edit/repos" target="_blank">
        View on production
      </a>

      { props.repositories.size === 0 ?

        <div className="empty">No repositories!</div> :

        props.repositories.map((repo, name) => 
          <Repository
            key={ name }
            { ...repo.toJS() }
            remove={ () => props.removeRepository(name) }
            change={ changeRepository }
            usedNames={ usedNames }
          />
        ).toArray()
      }

      <RepositoryAdd
        submit={ (name, path) => props.addRepository(name, path) }
        usedNames={ usedNames }
      />

    </div>
  );

}


export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Config);


