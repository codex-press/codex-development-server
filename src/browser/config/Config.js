import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux'

import {
  addRepository,
  removeRepository,
  changeRepository,
  setTokenStatus,
  impersonate,
  changeConfig,
  reload,
  navigate,
} from '../actions';

import AccountDetails from './AccountDetails';
import Repository from './Repository';
import RepositoryAdd from './RepositoryAdd';
import Impersonate from './Impersonate'; 

import './Config.less';


const mapStateToProps = state => {
  return {
    visible: state.getIn(['ui', 'modal']) === 'config',
    config: state.get('config'),
    ui: state.get('ui'),
    repositories: state.get('repositories'),
    domains: state.get('domains'),
    account: state.get('account'),
    users: state.get('users'),
  }
}


const mapDispatchToProps = {
  addRepository,
  removeRepository,
  changeRepository,
  setTokenStatus,
  impersonate,
  changeConfig,
  reload,
  navigate,
}


export function Config(props) {

  if (!props.visible)
    return null;

  const usedNames = props.repositories.toArray().map(a => a.get('name'));

  const changeRepository = name => (newName, path) => {
    props.changeRepository(name, newName, path) 
  }

  const changeCSP = e => {
    props.changeConfig('disable_csp', e.target.checked)
    props.reload();
  }

  const changeDomain = async e => {
    let value = e.target.value === "None" ? '' : e.target.value;
    await props.changeConfig('domain', value)
    props.navigate('/');
  }

  return (
    <div className="Config">

      <h2>Codex Press Develepment Server</h2>

      <Impersonate
        account={ props.account }
        users={ props.users }
        impersonate={ props.impersonate }
      />

      <AccountDetails
        account={ props.account }
        change={ () => props.setTokenStatus('changing') }
      />

      <label className="for-checkbox">
        <input
          type="checkbox"
          checked={ !!props.config.get('disable_csp') }
          onChange={ changeCSP }
        />
        Disable the Content Security Policy {}
        <a target="_blank" href="https://codex.press/docs/dev-server#csp">
          (help)
        </a>
      </label>

      { props.domains.size > 0 &&
        <label>
          Domain: {}
          <select
            value={ props.config.get('domain') }
            onChange={ changeDomain }>
            <option>None</option>
            { props.domains.map(d =>
              <option key={ d.get('id') }>
                { d.get('name') }
              </option>)
            }
          </select>
        </label>
      }

      <h3>Repositories</h3>

      <a href="https://codex.press/edit/repos" target="_blank">
        View on production
      </a>

      { props.repositories.size === 0 ?

        <div className="empty">No repositories!</div> :

        props.repositories.map((repo, name) => 
          <Repository
            key={ name }
            { ...repo.toJS() }
            remove={ () => props.removeRepository(name) }
            change={ changeRepository(name) }
            usedNames={ usedNames }
          />
        ).toArray()
      }

      <RepositoryAdd
        submit={ (name, path) => props.addRepository(name, path) }
        usedNames={ usedNames }
      />

    </div>
  );

}


export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Config);


