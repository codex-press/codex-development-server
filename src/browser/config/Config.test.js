import React from 'react'
import { shallow } from 'enzyme'
import { Map, fromJS } from 'immutable'

import { Config } from './Config'

var repositories = fromJS({
  app: {
    name: 'app',
    path: '/app',
  }
});


test('renders Config', () => {
  const c = shallow(
    <Config
      visible={ true }
      config={ Map({}) }
      repositories={ repositories }
      domains={ Map({}) }
      account={ Map({}) }
      users={ Map({}) }
    />
  )

  expect(c).toMatchSnapshot()
});


