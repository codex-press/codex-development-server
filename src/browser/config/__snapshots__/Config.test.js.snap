// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Config 1`] = `
<div
  className="Config"
>
  <h2>
    Codex Press Develepment Server
  </h2>
  <Impersonate
    account={
      Immutable.Map {
      }
    }
    users={
      Immutable.Map {
      }
    }
  />
  <AccountDetails
    account={
      Immutable.Map {
      }
    }
    change={[Function]}
  />
  <label
    className="for-checkbox"
  >
    <input
      checked={false}
      onChange={[Function]}
      type="checkbox"
    />
    Disable the Content Security Policy 
    <a
      href="https://codex.press/docs/dev-server#csp"
      target="_blank"
    >
      (help)
    </a>
  </label>
  <h3>
    Repositories
  </h3>
  <a
    href="https://codex.press/edit/repos"
    target="_blank"
  >
    View on production
  </a>
  <Repository
    change={[Function]}
    name="app"
    path="/app"
    remove={[Function]}
    usedNames={
      Array [
        "app",
      ]
    }
  />
  <RepositoryAdd
    submit={[Function]}
    usedNames={
      Array [
        "app",
      ]
    }
  />
</div>
`;
